### PR TITLE
Avoid rounding errors in _get_total_norm for a dTensor by using torch Dynamo

### LINKED ISSUE #149768: [FSDP2][DTensor] numeric bug for DTensor + python float in gradient clipping
### 🐛 Describe the bug

```
python3.12/site-packages/torch/nn/utils/clip_grad.py", line 155, in _clip_grads_with_norm_
[rank1]:     clip_coef = max_norm / (total_norm + 1e-6)
```

for DTensor + 1e-6, each rank adds 1e-6 to local tensor, instead of adding once

this was reported by others as well

### Versions

na

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @zhaojuanmao @mrshenli @rohan-varma @chauhang @mori360 @kwen2501 @c-p-i-o

### PR DESCRIPTION
**Pull request by RishabhSaini**
When a _NormPartial is Reduced, rounding errors can cause the resulting Tensor to have inconsistent results.
Example:
```
>>> import torch
>>> print(f"{(torch.linalg.vector_norm(torch.tensor([1.0, 1.0]))**2.0 + torch.linalg.vector_norm(torch.tensor([2.0, 2.0]))**2.0)**(1/2):.10f}")
3.1622774601
>>> print(f"{torch.linalg.vector_norm(torch.tensor([[1.0, 1.0], [2.0, 2.0]])):.10f}")
3.1622776985
```

Since the `_clip_grads_with_norm_` does: `clip_coef = max_norm / (total_norm + 1e-6)`, the results have larger rounding errors

Closes: #149768 

See alternative implementation details here waiting for approval:
https://github.com/pytorch/pytorch/issues/149768#issuecomment-2831707319

Edit:
Working on modifying the existing VectorNorm Strategy to avoid rounding errors 

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @zhaojuanmao @mrshenli @rohan-varma @chauhang @mori360 @kwen2501 @c-p-i-o

### CODE REVIEW

**Review by albanD — CHANGES_REQUESTED** [review]
This kind of small errors is expected. See https://pytorch.org/docs/stable/notes/numerical_accuracy.html#batched-computations-or-slice-computations for details.

Also the linked issue seemed to be talking about a different problem of the epsilong being applied multiple times?

**Comment by RishabhSaini** [comment]
>Also the linked issue seemed to be talking about a different problem of the epsilong being applied multiple times?

The epsilon is not added multiple times. I have verified that. Scaling the number of processes or nodes does not incrementally affect the Reduced value. 
The only reason for the numerical inaccuracy is due to separate calculations of _PartialNorm and its power (_PartialNorm**norm_type). When the all_reduce sums them, these inaccuracies scale up

**Comment by albanD** [comment]
Ho thanks or clarifying !

I'm afraid it doesn't change the fact that according to the floating point number standard, both of these results are valid so this kind of discrepancy is expected :/

**Comment by RishabhSaini** [comment]
Agreed, should I begin creating a `_PartialPowerSum` to address this? I am sure this would help other efforts as well:
https://github.com/pytorch/pytorch/issues/28119#issuecomment-575255083

**Comment by albanD** [comment]
These change would be in the DTensor implementation of linalg.vector_norm right?

**Comment by RishabhSaini** [comment]
Yes
